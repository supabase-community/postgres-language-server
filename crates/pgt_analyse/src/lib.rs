mod analysed_file_context;
mod categories;
pub mod context;
mod filter;
pub mod macros;
pub mod options;
mod registry;
pub mod rule;

// Re-exported for use in the `declare_group` macro
pub use pgt_diagnostics::category_concat;

pub use crate::analysed_file_context::AnalysedFileContext;
pub use crate::categories::{
    ActionCategory, RefactorKind, RuleCategories, RuleCategoriesBuilder, RuleCategory,
    SUPPRESSION_ACTION_CATEGORY, SourceActionKind,
};
pub use crate::filter::{AnalysisFilter, GroupKey, RuleFilter, RuleKey};
pub use crate::options::{AnalyserOptions, AnalyserRules};
pub use crate::registry::{
    MetadataRegistry, RegistryRuleParams, RegistryVisitor, RuleRegistry, RuleRegistryBuilder,
};
pub use crate::rule::{
    GroupCategory, Rule, RuleDiagnostic, RuleGroup, RuleMeta, RuleMetadata, RuleSource,
};
