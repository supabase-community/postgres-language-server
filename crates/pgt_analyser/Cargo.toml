
[package]
authors.workspace    = true
categories.workspace = true
description          = "<DESCRIPTION>"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "pgt_analyser"
repository.workspace = true
version              = "0.0.0"

[dependencies]
pgt_analyse      = { workspace = true }
pgt_console      = { workspace = true }
pgt_diagnostics  = { workspace = true }
pgt_query_ext    = { workspace = true }
pgt_schema_cache = { workspace = true }
pgt_text_size    = { workspace = true }
serde            = { workspace = true }

[dev-dependencies]
insta           = { version = "1.42.1" }
pgt_diagnostics = { workspace = true }
pgt_test_macros = { workspace = true }
termcolor       = { workspace = true }
