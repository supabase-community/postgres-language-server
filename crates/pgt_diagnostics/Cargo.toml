[package]
authors.workspace    = true
categories.workspace = true
description          = "<DESCRIPTION>"
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
name                 = "pgt_diagnostics"
repository.workspace = true
version              = "0.0.0"


[dependencies]
backtrace                  = "0.3.74"
bpaf                       = { workspace = true }
enumflags2                 = { workspace = true }
oxc_resolver               = { workspace = true }
pgt_console                = { workspace = true, features = ["serde"] }
pgt_diagnostics_categories = { workspace = true, features = ["serde"] }
pgt_diagnostics_macros     = { workspace = true }
pgt_text_edit              = { workspace = true, features = ["serde"] }
pgt_text_size.workspace    = true
schemars                   = { workspace = true, optional = true }
serde                      = { workspace = true, features = ["derive"] }
serde_json                 = { workspace = true }
termcolor                  = { workspace = true }
unicode-width              = { workspace = true }

[features]
schema = ["dep:schemars", "pgt_text_edit/schema", "pgt_diagnostics_categories/schema", "pgt_console/schema"]

[dev-dependencies]

[lib]
doctest = false
