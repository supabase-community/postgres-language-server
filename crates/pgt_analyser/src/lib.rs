use std::{ops::Deref, sync::LazyLock};

use pgt_analyse::{
    AnalysedFileContext, AnalyserOptions, AnalysisFilter, MetadataRegistry, RegistryRuleParams,
    RuleDiagnostic, RuleRegistry,
};
pub use registry::visit_registry;

mod lint;
pub mod options;
mod registry;

pub static METADATA: LazyLock<MetadataRegistry> = LazyLock::new(|| {
    let mut metadata = MetadataRegistry::default();
    visit_registry(&mut metadata);
    metadata
});

/// Main entry point to the analyser.
pub struct Analyser<'a> {
    /// Holds the metadata for all the rules statically known to the analyser
    /// we need this later when we add suppression support
    #[allow(dead_code)]
    metadata: &'a MetadataRegistry,

    /// Holds all rule options
    options: &'a AnalyserOptions,

    /// Holds all rules
    registry: RuleRegistry,
}

#[derive(Debug)]
pub struct AnalysableStatement {
    pub root: pgt_query_ext::NodeEnum,
    pub range: pgt_text_size::TextRange,
}

pub struct AnalyserParams<'a> {
    pub stmts: Vec<AnalysableStatement>,
    pub schema_cache: Option<&'a pgt_schema_cache::SchemaCache>,
}

pub struct AnalyserConfig<'a> {
    pub options: &'a AnalyserOptions,
    pub filter: AnalysisFilter<'a>,
}

impl<'a> Analyser<'a> {
    pub fn new(conf: AnalyserConfig<'a>) -> Self {
        let mut builder = RuleRegistry::builder(&conf.filter);
        visit_registry(&mut builder);
        let registry = builder.build();

        Self {
            metadata: METADATA.deref(),
            registry,
            options: conf.options,
        }
    }

    pub fn run(&self, params: AnalyserParams) -> Vec<RuleDiagnostic> {
        let mut diagnostics = vec![];

        let mut file_context = AnalysedFileContext::default();

        for stmt in params.stmts {
            let rule_params = RegistryRuleParams {
                root: &stmt.root,
                options: self.options,
                analysed_file_context: &file_context,
                schema_cache: params.schema_cache,
            };

            diagnostics.extend(
                self.registry
                    .rules
                    .iter()
                    .flat_map(|rule| (rule.run)(&rule_params)),
            );

            file_context.update_from(&stmt.root);
        }

        diagnostics
    }
}

#[cfg(test)]
mod tests {
    use core::slice;

    use pgt_analyse::{AnalyserOptions, AnalysisFilter, RuleFilter};
    use pgt_console::{
        Markup,
        fmt::{Formatter, Termcolor},
        markup,
    };
    use pgt_diagnostics::PrintDiagnostic;
    use pgt_text_size::TextRange;
    use termcolor::NoColor;

    use crate::{AnalysableStatement, Analyser};

    #[ignore]
    #[test]
    fn debug_test() {
        fn markup_to_string(markup: Markup) -> String {
            let mut buffer = Vec::new();
            let mut write = Termcolor(NoColor::new(&mut buffer));
            let mut fmt = Formatter::new(&mut write);
            fmt.write_markup(markup).unwrap();

            String::from_utf8(buffer).unwrap()
        }

        const SQL: &str = r#"alter table test drop column id;"#;
        let rule_filter = RuleFilter::Rule("safety", "banDropColumn");

        let filter = AnalysisFilter {
            enabled_rules: Some(slice::from_ref(&rule_filter)),
            ..Default::default()
        };

        let ast = pgt_query_ext::parse(SQL).expect("failed to parse SQL");
        let range = TextRange::new(0.into(), u32::try_from(SQL.len()).unwrap().into());

        let options = AnalyserOptions::default();

        let analyser = Analyser::new(crate::AnalyserConfig {
            options: &options,
            filter,
        });

        let results = analyser.run(crate::AnalyserParams {
            stmts: vec![AnalysableStatement { root: ast, range }],
            schema_cache: None,
        });

        println!("*******************");
        for result in &results {
            let text = markup_to_string(markup! {
                {PrintDiagnostic::simple(result)}
            });
            eprintln!("{}", text);
        }
        println!("*******************");

        // assert_eq!(results, vec![]);
    }
}
