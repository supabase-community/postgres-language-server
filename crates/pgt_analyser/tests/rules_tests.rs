use core::slice;
use std::{collections::HashMap, fmt::Write, fs::read_to_string, path::Path};

use pgt_analyse::{AnalyserOptions, AnalysisFilter, RuleDiagnostic, RuleFilter};
use pgt_analyser::{AnalysableStatement, Analyser, AnalyserConfig, AnalyserParams};
use pgt_console::StdDisplay;
use pgt_diagnostics::PrintDiagnostic;

pgt_test_macros::gen_tests! {
  "tests/specs/**/*.sql",
  crate::rule_test
}

fn rule_test(full_path: &'static str, _: &str, _: &str) {
    let input_file = Path::new(full_path);

    let (group, rule, fname) = parse_test_path(input_file);

    let rule_filter = RuleFilter::Rule(group.as_str(), rule.as_str());
    let filter = AnalysisFilter {
        enabled_rules: Some(slice::from_ref(&rule_filter)),
        ..Default::default()
    };

    let query =
        read_to_string(full_path).unwrap_or_else(|_| panic!("Failed to read file: {full_path} "));

    let options = AnalyserOptions::default();
    let analyser = Analyser::new(AnalyserConfig {
        options: &options,
        filter,
    });

    let split = pgt_statement_splitter::split(&query);

    let stmts = split
        .ranges
        .iter()
        .map(|r| {
            let text = &query[*r];
            let ast = pgt_query::parse(text).expect("failed to parse SQL");

            AnalysableStatement {
                root: ast.into_root().expect("Failed to convert AST to root node"),
                range: *r,
            }
        })
        .collect::<Vec<_>>();

    let results = analyser.run(AnalyserParams {
        stmts,
        schema_cache: None,
    });

    let mut snapshot = String::new();
    write_snapshot(&mut snapshot, query.as_str(), results.as_slice());

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => input_file.parent().unwrap(),
    }, {
        insta::assert_snapshot!(fname, snapshot);
    });

    let expectation = Expectation::from_file(&query);
    expectation.assert(results.as_slice());
}

fn parse_test_path(path: &Path) -> (String, String, String) {
    let mut comps: Vec<&str> = path
        .components()
        .map(|c| c.as_os_str().to_str().unwrap())
        .collect();

    let fname = comps.pop().unwrap();
    let rule = comps.pop().unwrap();
    let group = comps.pop().unwrap();

    (group.into(), rule.into(), fname.into())
}

fn write_snapshot(snapshot: &mut String, query: &str, diagnostics: &[RuleDiagnostic]) {
    writeln!(snapshot, "# Input").unwrap();
    writeln!(snapshot, "```").unwrap();
    writeln!(snapshot, "{query}").unwrap();
    writeln!(snapshot, "```").unwrap();
    writeln!(snapshot).unwrap();

    if !diagnostics.is_empty() {
        writeln!(snapshot, "# Diagnostics").unwrap();
        for diagnostic in diagnostics {
            let printer = PrintDiagnostic::simple(diagnostic);

            writeln!(snapshot, "{}", StdDisplay(printer)).unwrap();
            writeln!(snapshot).unwrap();
        }
    }
}

enum Expectation {
    NoDiagnostics,
    Diagnostics(Vec<(String, usize)>),
}

impl Expectation {
    fn from_file(content: &str) -> Self {
        let mut multiple_of: HashMap<&str, i32> = HashMap::new();
        for line in content.lines() {
            if line.contains("expect_no_diagnostics") {
                if !multiple_of.is_empty() {
                    panic!(
                        "Cannot use both `expect_no_diagnostics` and `expect_` in the same test"
                    );
                }
                return Self::NoDiagnostics;
            }

            if line.contains("expect_") && !line.contains("expect_no_diagnostics") {
                let kind = line
                    .splitn(3, "_")
                    .last()
                    .expect("Use pattern: `-- expect_<category>`")
                    .trim();

                *multiple_of.entry(kind).or_insert(0) += 1;
            }
        }

        if !multiple_of.is_empty() {
            return Self::Diagnostics(
                multiple_of
                    .into_iter()
                    .map(|(k, v)| (k.into(), v as usize))
                    .collect(),
            );
        }

        panic!(
            "No expectation found in the test file. Use `-- expect_no_diagnostics` or `-- expect_<category>`"
        );
    }

    fn assert(&self, diagnostics: &[RuleDiagnostic]) {
        match self {
            Self::NoDiagnostics => {
                if !diagnostics.is_empty() {
                    panic!("This test should not have any diagnostics.");
                }
            }
            Self::Diagnostics(expected) => {
                let mut counts: HashMap<&str, usize> = HashMap::new();
                for diag in diagnostics {
                    *counts.entry(diag.get_category_name()).or_insert(0) += 1;
                }

                for (kind, expected_count) in expected {
                    let actual_count = counts.get(kind.as_str()).copied().unwrap_or(0);
                    if actual_count != *expected_count {
                        panic!(
                            "Expected {expected_count} diagnostics of kind `{kind}`, but found {actual_count}."
                        );
                    }
                }
            }
        }
    }
}
