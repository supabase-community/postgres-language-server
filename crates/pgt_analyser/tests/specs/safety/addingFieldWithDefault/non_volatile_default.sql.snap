---
source: crates/pgt_analyser/tests/rules_tests.rs
expression: snapshot
snapshot_kind: text
---
# Input
```
-- Test non-volatile default values (should be safe in PG 11+, but we are passing no PG version info in the tests)
-- expect_lint/safety/addingFieldWithDefault
ALTER TABLE users ADD COLUMN status text DEFAULT 'active';

```

# Diagnostics
lint/safety/addingFieldWithDefault ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × Adding a column with a DEFAULT value causes a table rewrite.
  
  i This operation requires an ACCESS EXCLUSIVE lock and rewrites the entire table.
  
  i Add the column without a default, then set the default in a separate statement.
